test/Microsoft.AspNet.Razor.Runtime.Test/TagHelpers/TagHelperAttributeListTest.cs

```diff
@@ -0,0 +1,661 @@
+                    { new[] { B }, "Set", set, new[] { B, set } },
+                    { new[] { B }, "Set", "Set Value", new[] { B, set } },
+                    { new[] { A, B }, "Set", set, new[] { A, B, set } },
+                    { new[] { A, B }, "Set", "Set Value", new[] { A, B, set } },
+
+                    // Multiple elements same name
+                    { new[] { A, B, A2, C }, "AName", set, new[] { set, B, C } },
+                    { new[] { A, B, A2, C }, "aname", set, new[] { set, B, C } },
+                    { new[] { A, B, A2, C }, "aname", "av", new[] { new TagHelperAttribute("aname", "av"), B, C } },
+                    { new[] { B, A2, A }, "aname", set, new[] { B, set } },
+                    { new[] { B, A2, A }, "aname", "av", new[] { B, new TagHelperAttribute("aname", "av") } },
+                    { new[] { B, A2, A, C }, "AName", set, new[] { B, set, C } },
+                    { new[] { B, A2, A, C }, "AName", "av", new[] { B, new TagHelperAttribute("AName", "av"), C } },
+                    { new[] { A, A3 }, "AName", set, new[] { set } },
+                    { new[] { A, A3 }, "AName", "av", new[] { new TagHelperAttribute("AName", "av") } },
+                    { new[] { A3, A }, "aname", set, new[] { set } },
+                    { new[] { A3, A }, "aname", "av", new[] { new TagHelperAttribute("aname", "av") } },
+                    { new[] { A, A2, A3 }, "AName", set, new[] { set } },
+                    { new[] { A, A2, A3 }, "AName", "av", new[] { new TagHelperAttribute("AName", "av") } },
+                    { new[] { C, B, A3, A }, "AName", set, new[] { C, B, set } },
+                    { new[] { C, B, A3, A }, "AName", "av", new[] { C, B, new TagHelperAttribute("AName", "av") } },
+                    { new[] { A, A2, A3 }, "Set", set, new[] { A, A2, A3, set } },
+                    { new[] { A, A2, A3 }, "Set", "Set Value", new[] { A, A2, A3, set } },
+                    { new[] { A, A2, A3, B, C }, "Set", set, new[] { A, A2, A3, B, C, set } },
+                    { new[] { A, A2, A3, B, C }, "Set", "Set Value", new[] { A, A2, A3, B, C, set } },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(StringIndexerSetData))]
+        public void StringIndexer_SetsAttributeAtExpectedLocation(
+            IEnumerable<TagHelperAttribute> initialAttributes,
+            string keyToSet,
+            TagHelperAttribute setValue,
+            IEnumerable<TagHelperAttribute> expectedAttributes)
+        {
+            // Arrange
+            var attributes = new TagHelperAttributeList(initialAttributes);
+
+            // Act
+            attributes[keyToSet] = setValue;
+
+            // Assert
+            Assert.Equal(expectedAttributes, attributes, CaseSensitiveTagHelperAttributeComparer.Default);
+        }
+
+        [Fact]
+        public void StringIndexer_SetsAttributeThrowsIfInvalidIndex()
```

Comment:
no `SetsAttribute` in cases that `throw`.  this should be `StringIndexer_Setter_ThrowsIfIndexInvalid`.  similar for next test (`StringIndexer_Setter_ThrowsIfNamesAreInconsistent`)